backend/git_real/activity_log_creators.py

```diff
@@ -42,20 +44,34 @@ def log_pr_closed(pull_request):
 def log_pr_opened(pull_request):
 
     event_type, _ = EventType.objects.get_or_create(name=PR_OPENED)
-
     match = LogEntry.objects.filter(
         actor_user=pull_request.user,
         effected_user=pull_request.user,
-        object_1=pull_request,
         event_type=event_type,
         timestamp__gte=timezone.now() - timezone.timedelta(minutes=2),
-    ).first()
-
+        object_1_content_type = ContentType.objects.get(model='pullrequest')
+        ).filter(
+            object_1_id = pull_request.id
+        ).first()
+    
     if match == None:
+        recruit_projects = curriculum_tracking.models.RecruitProject.objects.filter(
+                repository=pull_request.repository,
+            )
+        recruit_project_start_time = None
+        for recruit_project in recruit_projects[::-1]:
+            recruit_project_start_time = recruit_project.start_time
+            if not recruit_project_start_time:
+                continue
+            if pull_request.created_at >= recruit_project_start_time:
+                break
```

Comment:
This doesn't really make sense to me, I'm thinking about some of those Cloudbnb projects, where there are multiple projects on one repo and I've looked at a learners PRS where there are 2 PRs open, one for Project part 1 and one for project part 2. So this would mean that the date/time a PR is created could be after the start time of both parts of the project so how would we know we are linking the correct project to the opened PR?
(if you weren't allowed to start the next part of the project until the previous part was finished then this would work but not all projects are setup like that....)

Reply:
We will need to guess a little bit here. 

```
def get_card(): 
    all_cards = find all AgileCards that could be associated with that PR (the repo->project->card)
    for status in [REVIEW_FEEDBACK, IN_PROGRESS, IN_REVIEW, COMPLETE,BACKLOG,BLOCKED]:  # this order matters
         filtered_cards = all_cards.filter(status=status)
         if len(filtered_cards):
               order the cards by time spent in the column 
               return oldest? Or maybe this is different per column. 
```

The reason for the ordering is:
- REVIEW_FEEDBACK: it's more important to address feedback than work on new code, this assumes the learners are being good
- IN_PROGRESS: this is the next column the learner should be working on. After this point it gets weird
- IN_REVIEW: maybe a learner has a conversation with a reviewer and pushes more code or something. Who knows.

The rest of the columns are included so that if anything truly weird happens at least the activity will be associated with something. The rest of the columns are ordered according to weirdness, in ascending order.

And then we will just need to keep an eye on things. This log might tell us a bit about how the learners behave when it comes to prioritising code.

A completely different alternative would be to just associate the event with the repo, we would lose information though. So it's either guess a little, or lose resolution as far as I can see.